cirq-core/cirq/ops/raw_types.py

```diff
@@ -375,6 +377,49 @@ def _rmul_with_qubits(self, qubits: Tuple['cirq.Qid', ...], other):
     def _json_dict_(self) -> Dict[str, Any]:
         return protocols.obj_to_dict_helper(self, attribute_names=[])
 
+    def on_each(self, *targets: Union[Qid, Iterable[Any]]) -> List['Operation']:
+        """Returns a list of operations applying the gate to all targets.
+
+        Args:
+            *targets: The qubits to apply this gate to. This can be provided as
+            nested iterables, but each leaf level must contain an exact
+            multiple of `num_qubits` qubits.
+
+        Returns:
+            Operations applying this gate to the target qubits.
+
+        Raises:
+            ValueError if targets are not instances of Qid or Iterable[Qid].
+            ValueError if the gate qubit number is incompatible.
+        """
+        operations: List['Operation'] = []
+        while any(targets):
```

Comment:
what's the check here? making sure some of the targets aren't empty lists? are we accepting empty lists as an acceptable input? does it make sense for `gate.on_each(False, None)` to return successfully?

Reply:
It's a while loop, and `targets` is updated in the loop. For an n-qubit gate it pops off the first n values each iteration of the loop (or the first single value if it's an iterator).

I tried to make it so that it was both strict for multi-qubit gates *and* consistent between single-qubit and multi-qubit gates. So single-qubit gates iterate through an iterator tree that has single qubits, two-qubit gates iterate through a tree that has pairs, etc. It ensures strictness, in that you have to give a two-qubit gate targets that occur in pairs, but it allows the same flexibility that we give to single-qubit gates in that those targets can occur in nested iterators or flat lists.

If we'd rather make it such that multi-qubit gates accept only `Iterable[Tuple[Qid, ...]]` and don't allow nesting, I can change that.

Reply:
I'm pretty sure the original issue notes that we should not allow `CNOT.on_each(a, b)`

Reply:
also pretty sure that `CNOT.on_each(False, None)` as currently implemented would silently return `[]`, which is surprising

Reply:
> which is surprising

Ah I missed that. The intention was `while len(targets) != 0`. But I'll go ahead and rewrite this such that multi-qubit gates only accept iterables with no nesting.